scaleway/resource_ssh_key.go

```diff
@@ -0,0 +1,129 @@
+package scaleway
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	api "github.com/nicolai86/scaleway-sdk"
+	"golang.org/x/crypto/ssh"
+)
+
+func resourceScalewaySSHKey() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceScalewaySSHKeyCreate,
+		Read:   resourceScalewaySSHKeyRead,
+		Delete: resourceScalewaySSHKeyDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"key": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The ssh key",
+			},
+		},
+	}
+}
+
+func getSSHKeyFingerprint(key []byte) (string, error) {
+	pubkey, _, _, _, err := ssh.ParseAuthorizedKey(key)
+	if err != nil {
+		return "", err
+	}
+	return ssh.FingerprintLegacyMD5(pubkey), nil
+}
+
+func resourceScalewaySSHKeyCreate(d *schema.ResourceData, m interface{}) error {
+	scaleway := m.(*Client).scaleway
+
+	fingerprint, err := getSSHKeyFingerprint([]byte(d.Get("key").(string)))
+	if err != nil {
+		return err
+	}
+
+	user, err := scaleway.GetUser()
+	if err != nil {
+		return err
+	}
+
+	keys := []api.KeyDefinition{}
+	exists := false
+	for _, key := range user.SSHPublicKeys {
+		exists = exists || key.Key == d.Get("key").(string)
+		keys = append(keys, api.KeyDefinition{
+			Key: key.Key,
+		})
+	}
+
+	// remote already contains the key, nothing to do
+	if exists {
+		d.SetId(fingerprint)
+		return nil
+	}
+
+	mu.Lock()
```

Comment:
do you want to move this lock up / re-retrieve the list post-lock, in-case a secondary key gets added in the meantime?

Reply:
done!